pkg/collector/scoped_distinct_string.go

```diff
@@ -4,31 +4,44 @@ import (
 	"sync"
 )
 
+const IntrinsicScope = "intrinsic"
+
 type ScopedDistinctString struct {
-	cols        map[string]*DistinctString
-	newCol      func(int) *DistinctString
-	maxLen      int
-	curLen      int
-	limExceeded bool
-	diffEnabled bool
-	mtx         sync.Mutex
+	cols            map[string]*DistinctString
+	newCol          func(int, uint32) *DistinctString
+	maxDataSize     int
+	currDataSize    int
+	limExceeded     bool
+	diffEnabled     bool
+	maxTagsPerScope uint32
+	mtx             sync.Mutex
 }
 
-func NewScopedDistinctString(maxDataSize int) *ScopedDistinctString {
+// NewScopedDistinctString collects the tags per scope
+// MaxDataSize is calculated as the total length of the recorded strings.
+// MaxTagsPerScope controls how many tags can be added per scope. The intrinsic scope is unbounded.
+// For ease of use, maxDataSize=0 and maxTagsPerScope=0 are interpreted as unlimited.
+func NewScopedDistinctString(maxDataSize int, maxTagsPerScope uint32) *ScopedDistinctString {
 	return &ScopedDistinctString{
-		cols:        map[string]*DistinctString{},
-		newCol:      NewDistinctString,
-		maxLen:      maxDataSize,
-		diffEnabled: false,
+		cols:            map[string]*DistinctString{},
+		newCol:          NewDistinctString,
+		maxDataSize:     maxDataSize,
+		diffEnabled:     false,
+		maxTagsPerScope: maxTagsPerScope,
 	}
 }
 
-func NewScopedDistinctStringWithDiff(maxDataSize int) *ScopedDistinctString {
+// NewScopedDistinctStringWithDiff collects the tags per scope with diff
+// MaxDataSize is calculated as the total length of the recorded strings.
+// MaxTagsPerScope controls how many tags can be added per scope. The intrinsic scope is unbounded.
+// For ease of use, maxDataSize=0 and maxTagsPerScope=0 are interpreted as unlimited.
+func NewScopedDistinctStringWithDiff(maxDataSize int, maxTagsPerScope uint32) *ScopedDistinctString {
```

Comment:
I thought about that as well, if we split the limit between tags (in the same way the maxSize is working right now) we could end with an incomplete list of span tags and none of the resource tags. The way this limit works now cannot be worse than the current implementation but in the worst-case scenario it will end with an exhaustive search.

I'm not opposed to keep the search deterministic by scope.

Reply:
The best pattern I can come up with for a tag names limit is to stop searching if ANY scope exceeds the limit instead of ALL scopes.

Reply:
Agree I think it's a good compromise